### PR TITLE
Add more test cases to test libnetwork API 

### PR DESCRIPTION
**Pull request by mrjana**
- Added more testcases for libnetwork API testing
- Added new error types for all of libnetwork errors
- Added a test binary to test README.md code
- Fixed a bug in bridge driver where when the bridge already exists
  the bridgeInterface.bridgeIPv4 is not getting initialized properly

### CODE REVIEW

File: cmd/readme_test/readme.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package main
```
Comment (mavenugo) [inline cmd/readme_test/readme.go]: Is there a reason why you didnt modify `cmd/test/main.go` to reflect the readme ?
I dont see a need to have 2 different sample code both are for the same purpose.

File: drivers/bridge/setup_verify.go [hunk 2]
```diff
@@ -29,6 +29,12 @@ func setupVerifyConfiguredAddresses(config *Configuration, i *bridgeInterface) e
 		return fmt.Errorf("Bridge IPv6 addresses do not match the expected bridge configuration %s", bridgeIPv6)
 	}
 
+	// By this time we have either configured a new bridge with an IP address
+	// or made sure an existing bridge's IP matches the configuration
+	// Now is the time to cache these states in the bridgeInterface.
+	i.bridgeIPv4 = addrv4.IPNet
+	i.bridgeIPv6 = bridgeIPv6
```
Comment (mavenugo) [inline drivers/bridge/setup_verify.go]: When a method is named something like `VerifyConfiguredAddresses` in setup_verify.go, I would assume that the method just performs verification. Updates of any sort makes this method misleading.

File: error.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+// UnknownEndpointError is returned when libnetwork could not find in it's database
+// an endpoint with the same name and id.
+type UnknownEndpointError struct {
+	name string
+	id   string
+}
+
+func (uee *UnknownEndpointError) Error() string {
+	return fmt.Sprintf("unknown endpoint %s id %s", uee.name, uee.id)
+}
```
Comment (mavenugo) [inline error.go]: :+1: Thanks for taking care of this. Hugely beneficial for error handling.

File: libnetwork_test.go [hunk 4]
```diff
@@ -73,3 +214,70 @@ func TestSimplebridge(t *testing.T) {
...
+
+	if _, ok := err.(*libnetwork.UnknownEndpointError); !ok {
+		t.Fatalf("Did not fail with expected error. Actual error: %v", err)
+	}
+
+	// Done testing. Now cleanup
+	if err := network.Delete(); err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (mavenugo) [inline libnetwork_test.go]: :+1:  Thanks for the diligence.

File: cmd/readme_test/readme.go (hunk 1 above)
Comment (mrjana) [inline cmd/readme_test/readme.go]: The reason cmd/test/main.go will at some point become become a full fledged test application with the api becomes complete. While the readme test I would like keep it uncluttered just to test the `README.md` code snippet

File: drivers/bridge/setup_verify.go (hunk 2 above)
Comment (mrjana) [inline drivers/bridge/setup_verify.go]: Ok. I will change the function name to `setupVerifyAndReconcile`

File: libnetwork_test.go (hunk 4 above)
Comment (mrjana) [inline libnetwork_test.go]: I don't think we need that there. `README.md` code snippet is like a quick start quide which quickly brings the user to speed on how to use the APIs. If it becomes too comprehensive it becomes a reference manual :-)

**Comment by mrjana** [comment]
Updated diffs with @mavenugo comments

**Comment by aboch** [comment]
LGTM

File: cmd/readme_test/readme.go (hunk 1 above)
Comment (mavenugo) [inline cmd/readme_test/readme.go]: okay. since it is not impacting the code in anyway, and it serves a specific purpose, am fine with that.

**Pull request merged**